src/core/localize.rkt

```diff
@@ -139,77 +134,215 @@
   (define nodes (batch-nodes expr-batch))
   (define roots (batch-roots expr-batch))
 
-  ; TODO don't ignore the status code from make-real-compiler in eval-progs-real
   (define subexprs-fn (eval-progs-real (map prog->spec exprs-list) ctx-list))
-  (define actual-value-fn (compile-progs exprs-list ctx))
 
   (define errs
     (for/vector #:length (vector-length roots)
                 ([node (in-vector roots)])
       (make-vector (pcontext-length (*pcontext*)))))
 
+  (for ([(pt ex) (in-pcontext (*pcontext*))]
+        [pt-idx (in-naturals)])
+    (define exacts (list->vector (apply subexprs-fn pt)))
+    (for ([expr (in-list exprs-list)]
+          [root (in-vector roots)]
+          [exact (in-vector exacts)]
+          [expr-idx (in-naturals)])
+      (define err
+        (match (vector-ref nodes root)
+          [(? literal?) 1]
+          [(? variable?) 1]
+          [(approx _ impl)
+           (define repr (repr-of expr ctx))
+           (ulp-difference exact (vector-ref exacts (vector-member impl roots)) repr)]
+          [`(if ,c ,ift ,iff) 1]
+          [(list f args ...)
+           (define repr (impl-info f 'otype))
+           (define argapprox
+             (for/list ([idx (in-list args)])
+               (vector-ref exacts (vector-member idx roots)))) ; arg's index mapping to exact
+           (define approx (apply (impl-info f 'fl) argapprox))
+           (ulp-difference exact approx repr)]))
+      (vector-set! (vector-ref errs expr-idx) pt-idx err)))
+
+  (define n 0)
+  (for/list ([subexprs (in-list subexprss)])
+    (for*/hash ([subexpr (in-list subexprs)])
+      (begin0 (values subexpr (vector->list (vector-ref errs n)))
+        (set! n (add1 n))))))
+
+(define (check-for-invalid-exact input)
+  (match input
+    ['+inf.0 #t]
+    ['-inf.0 #t]
+    ['+nan.0 #t]
+    [value (or (bfnan? value) (boolean? value))]))
+
+; Compute local error or each sampled point at each node in `prog`.
+(define (compute-errors subexprss ctx)
+  (define exprs-list (append* subexprss)) ; unroll subexprss
+  (define spec-list (map prog->spec exprs-list))
+  (define ctx-list
+    (for/list ([subexpr (in-list exprs-list)])
+      (struct-copy context ctx [repr (repr-of subexpr ctx)])))
+  (define exact-var-name '__exact)
+  (define extended
+    (for/list ([ctx (in-list ctx-list)])
+      (context-extend ctx exact-var-name (context-repr ctx))))
+  (define compare-specs
+    (for/list ([spec (in-list spec-list)])
+      `(- ,spec ,exact-var-name)))
```

Comment:
I'm confused. The exact value for each spec is different. Don't you need a different variable name for each one?